compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -124,39 +125,85 @@ abstract class GenJSCode extends plugins.PluginComponent
     override def description: String = GenJSCode.this.description
     override def erasedTypes: Boolean = true
 
-    // Some state --------------------------------------------------------------
+    // Scoped state ------------------------------------------------------------
 
-    val currentClassSym          = new ScopedVar[Symbol]
-    val currentMethodSym         = new ScopedVar[Symbol]
-    val thisLocalVarIdent        = new ScopedVar[Option[js.Ident]](None)
-    val fakeTailJumpParamRepl    = new ScopedVar[(Symbol, Symbol)]((NoSymbol, NoSymbol))
-    val enclosingLabelDefParams  = new ScopedVar(Map.empty[Symbol, List[Symbol]])
-    val mutableLocalVars         = new ScopedVar[mutable.Set[Symbol]]
-    val mutatedLocalVars         = new ScopedVar[mutable.Set[Symbol]]
-    val unexpectedMutatedFields  = new ScopedVar[mutable.Set[Symbol]]
-    val paramAccessorLocals      = new ScopedVar(Map.empty[Symbol, js.ParamDef])
+    //// Per class body
+    val currentClassSym = new ScopedVar[Symbol]
+    private val unexpectedMutatedFields = new ScopedVar[mutable.Set[Symbol]]
 
-    var isModuleInitialized: Boolean = false // see genApply for super calls
+    private def currentClassType = encodeClassType(currentClassSym)
 
-    val countsOfReturnsToMatchEnd = mutable.Map.empty[Symbol, Int]
+    //// Per method body
+    private val currentMethodSym = new ScopedVar[Symbol]
+    private val thisLocalVarIdent = new ScopedVar[Option[js.Ident]]
+    private val fakeTailJumpParamRepl = new ScopedVar[(Symbol, Symbol)]
+    private val enclosingLabelDefParams = new ScopedVar[Map[Symbol, List[Symbol]]]
+    private val isModuleInitialized = new ScopedVar[VarBox[Boolean]]
+    private val undefinedDefaultParams = new ScopedVar[mutable.Set[Symbol]]
 
-    private def currentClassType = encodeClassType(currentClassSym)
+    //// For some method bodies
+    private val mutableLocalVars = new ScopedVar[mutable.Set[Symbol]]
+    private val mutatedLocalVars = new ScopedVar[mutable.Set[Symbol]]
+
+    //// For anonymous methods
+    // These have a default, since we always read them.
+    private val tryingToGenMethodAsJSFunction = new ScopedVar[Boolean](false)
+    private val paramAccessorLocals = new ScopedVar(Map.empty[Symbol, js.ParamDef])
 
-    val tryingToGenMethodAsJSFunction = new ScopedVar[Boolean](false)
-    class CancelGenMethodAsJSFunction(message: String)
+    private class CancelGenMethodAsJSFunction(message: String)
         extends Throwable(message) with scala.util.control.ControlThrowable
 
-    // Rewriting of anonymous function classes ---------------------------------
+    //// For matches
+    private val countsOfReturnsToMatchEnd = new ScopedVar[VarBox[Int]]
```

Comment:
I don't think dropping the `Map` is valid. In case of nested matches, or worse, in the case of a `match` in the *guard* of another `match`, I'm pretty sure it's possible to have jumps to a non-directly enclosing matchEnd. I think this should be a `ScopedVar[mutable.Map]`